# Pandas
Pandas is a high performance easy to use tool for data analysis in python programming.
Pandas provides two important data structures:
1. Series: A pandas series is a single column (Similar to lists).
2. Dataframes: Can be considered as relational table with rows representing an instance of an example and columns representing different variables

In [1]:
from __future__ import print_function
import pandas as pd
import matplotlib.pyplot as plt 
import numpy as np

## Series
A pandas series is a single column. Similar to lists

In [2]:
city_names = pd.Series(['New Delhi', 'Mumbai', 'New York', 'London'])
city_names

0    New Delhi
1       Mumbai
2     New York
3       London
dtype: object

In [3]:
population_in_M = pd.Series([18.7, 23.6, 22, 8.7])
population_in_M

0    18.7
1    23.6
2    22.0
3     8.7
dtype: float64

## Dataframe
Can be considered as relational table with rows representing an instance of an example and columns representing different variables

In [4]:
cities = pd.DataFrame({'City_names': city_names, 'Population_in_M' : population_in_M})
cities

City_names  Population_in_M
0  New Delhi             18.7
1     Mumbai             23.6
2   New York             22.0
3     London              8.7

In [5]:
#Using index parameter to change the index
cities = pd.DataFrame({'City_names': list(city_names), 'Population_in_M' : list(population_in_M)}, index=['I', 'II', 'III', 'IV'])
cities

City_names  Population_in_M
I    New Delhi             18.7
II      Mumbai             23.6
III   New York             22.0
IV      London              8.7

## Reading from csv

In [6]:
california_housing_dataframe = pd.read_csv("https://download.mlcc.google.com/mledu-datasets/california_housing_train.csv", sep=",")
#To see first few sample from the dataframe
california_housing_dataframe.head()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -114.31     34.19                15.0       5612.0          1283.0   
1    -114.47     34.40                19.0       7650.0          1901.0   
2    -114.56     33.69                17.0        720.0           174.0   
3    -114.57     33.64                14.0       1501.0           337.0   
4    -114.57     33.57                20.0       1454.0           326.0   

   population  households  median_income  median_house_value  
0      1015.0       472.0         1.4936             66900.0  
1      1129.0       463.0         1.8200             80100.0  
2       333.0       117.0         1.6509             85700.0  
3       515.0       226.0         3.1917             73400.0  
4       624.0       262.0         1.9250             65500.0

In [7]:
california_housing_dataframe.head(10)

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -114.31     34.19                15.0       5612.0          1283.0   
1    -114.47     34.40                19.0       7650.0          1901.0   
2    -114.56     33.69                17.0        720.0           174.0   
3    -114.57     33.64                14.0       1501.0           337.0   
4    -114.57     33.57                20.0       1454.0           326.0   
5    -114.58     33.63                29.0       1387.0           236.0   
6    -114.58     33.61                25.0       2907.0           680.0   
7    -114.59     34.83                41.0        812.0           168.0   
8    -114.59     33.61                34.0       4789.0          1175.0   
9    -114.60     34.83                46.0       1497.0           309.0   

   population  households  median_income  median_house_value  
0      1015.0       472.0         1.4936             66900.0  
1      1129.0       463.0         1.8200             80100.0  
2       333.0       117.0         1.6509             85700.0  
3       515.0       226.0         3.1917             73400.0  
4       624.0       262.0         1.9250             65500.0  
5       671.0       239.0         3.3438             74000.0  
6      1841.0       633.0         2.6768             82400.0  
7       375.0       158.0         1.7083             48500.0  
8      3134.0      1056.0         2.1782             58400.0  
9       787.0       271.0         2.1908             48100.0

In [8]:
#Gves the overall shpae of the entire dataframe
california_housing_dataframe.shape

(17000, 9)

In [9]:
#Getting statistical description of the data
california_housing_dataframe.describe()

longitude      latitude  housing_median_age   total_rooms  \
count  17000.000000  17000.000000        17000.000000  17000.000000   
mean    -119.562108     35.625225           28.589353   2643.664412   
std        2.005166      2.137340           12.586937   2179.947071   
min     -124.350000     32.540000            1.000000      2.000000   
25%     -121.790000     33.930000           18.000000   1462.000000   
50%     -118.490000     34.250000           29.000000   2127.000000   
75%     -118.000000     37.720000           37.000000   3151.250000   
max     -114.310000     41.950000           52.000000  37937.000000   

       total_bedrooms    population    households  median_income  \
count    17000.000000  17000.000000  17000.000000   17000.000000   
mean       539.410824   1429.573941    501.221941       3.883578   
std        421.499452   1147.852959    384.520841       1.908157   
min          1.000000      3.000000      1.000000       0.499900   
25%        297.000000    790.000000    282.000000       2.566375   
50%        434.000000   1167.000000    409.000000       3.544600   
75%        648.250000   1721.000000    605.250000       4.767000   
max       6445.000000  35682.000000   6082.000000      15.000100   

       median_house_value  
count        17000.000000  
mean        207300.912353  
std         115983.764387  
min          14999.000000  
25%         119400.000000  
50%         180400.000000  
75%         265000.000000  
max         500001.000000

## Indexes

In [10]:
cities

City_names  Population_in_M
I    New Delhi             18.7
II      Mumbai             23.6
III   New York             22.0
IV      London              8.7

In [11]:
cities.index

Index(['I', 'II', 'III', 'IV'], dtype='object')

In [12]:
california_housing_dataframe.index

RangeIndex(start=0, stop=17000, step=1)

In [13]:
cities['Initials'] = ['ND', 'M', 'NY', 'L']
cities
cities.reindex(np.random.permutation(cities.index))

City_names  Population_in_M Initials
IV      London              8.7        L
III   New York             22.0       NY
II      Mumbai             23.6        M
I    New Delhi             18.7       ND

In [14]:
#Setting and changing indexes
cities.set_index("Initials")

City_names  Population_in_M
Initials                            
ND        New Delhi             18.7
M            Mumbai             23.6
NY         New York             22.0
L            London              8.7

## Accessing data from Dataframe

In [15]:
#Accessing an entire column based on a variable/column name the python dictionary way
cities['Population_in_M']

I      18.7
II     23.6
III    22.0
IV      8.7
Name: Population_in_M, dtype: float64

In [16]:
type(cities['Population_in_M'])

pandas.core.series.Series

In [17]:
#Naive way
cities.City_names

I      New Delhi
II        Mumbai
III     New York
IV        London
Name: City_names, dtype: object

In [18]:
#Accessing particular cell from a column using index
cities['City_names'][2]

'New York'

In [19]:
type(cities['City_names'][2])

str

In [20]:
# Accessing a varuable for it's description
california_housing_dataframe.total_bedrooms.describe()

count    17000.000000
mean       539.410824
std        421.499452
min          1.000000
25%        297.000000
50%        434.000000
75%        648.250000
max       6445.000000
Name: total_bedrooms, dtype: float64

In [21]:
california_housing_dataframe.total_bedrooms.mean()

539.4108235294118

In [40]:
cities.City_names.describe()

count          4
unique         4
top       London
freq           1
Name: City_names, dtype: object

In [41]:
cities.City_names.unique()

array(['New Delhi', 'Mumbai', 'New York', 'London'], dtype=object)

In [42]:
cities.City_names.value_counts()

London       1
Mumbai       1
New York     1
New Delhi    1
Name: City_names, dtype: int64

### Indexing in Pandas

#### iloc operator: index-based selection

In [22]:
#Accessing a single row from dataframe
cities.iloc[1]

City_names         Mumbai
Population_in_M      23.6
Initials                M
Name: II, dtype: object

In [23]:
type(cities.iloc[1])

pandas.core.series.Series

In [24]:
cities.iloc[-1]

City_names         London
Population_in_M       8.7
Initials                L
Name: IV, dtype: object

In [25]:
california_housing_dataframe.iloc[25:29]

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
25    -115.32     32.82                34.0        591.0           139.0   
26    -115.37     32.82                30.0       1602.0           322.0   
27    -115.37     32.82                14.0       1276.0           270.0   
28    -115.37     32.81                32.0        741.0           191.0   

    population  households  median_income  median_house_value  
25       327.0        89.0         3.6528            100000.0  
26      1130.0       335.0         3.5735             71100.0  
27       867.0       261.0         1.9375             80900.0  
28       623.0       169.0         1.7604             68600.0

In [26]:
type(california_housing_dataframe.iloc[0:3])

pandas.core.frame.DataFrame

In [27]:
# Retrieving an entirecolumn using iloc
cities.iloc[:,0]

I      New Delhi
II        Mumbai
III     New York
IV        London
Name: City_names, dtype: object

In [28]:
# Fetching few elements from a column
# Fetching first 10 entries from total_rooms
california_housing_dataframe.iloc[:10, 3]

0    5612.0
1    7650.0
2     720.0
3    1501.0
4    1454.0
5    1387.0
6    2907.0
7     812.0
8    4789.0
9    1497.0
Name: total_rooms, dtype: float64

In [29]:
#Accessing a list of entries
cities.iloc[[1,3],0]

II    Mumbai
IV    London
Name: City_names, dtype: object

#### loc operator: label-based selection.

In [30]:
cities.loc['II', 'City_names']

'Mumbai'

In [31]:
california_housing_dataframe.loc[2:5, 'population']

2    333.0
3    515.0
4    624.0
5    671.0
Name: population, dtype: float64

In [32]:
california_housing_dataframe.loc[2:5, ['latitude', 'longitude']]

latitude  longitude
2     33.69    -114.56
3     33.64    -114.57
4     33.57    -114.57
5     33.63    -114.58

*iloc uses the Python stdlib indexing scheme, where the first element of the range is included and the last one excluded. So 0:10 will select entries 0,...,9. loc, meanwhile, indexes inclusively. So 0:10 will select entries 0,...,10.*

**Use *loc* when index are alpha-numeric or just words, and use *iloc* when indexes are only numeric**

### Conditional Selection

In [33]:
california_housing_dataframe.total_bedrooms > 500

0         True
1         True
2        False
3        False
4        False
         ...  
16995    False
16996     True
16997     True
16998     True
16999    False
Name: total_bedrooms, Length: 17000, dtype: bool

In [34]:
california_housing_dataframe.loc[california_housing_dataframe.total_bedrooms > 1000]

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0        -114.31     34.19                15.0       5612.0          1283.0   
1        -114.47     34.40                19.0       7650.0          1901.0   
8        -114.59     33.61                34.0       4789.0          1175.0   
115      -115.85     34.20                34.0       3868.0          1257.0   
133      -116.06     34.15                15.0      10377.0          2331.0   
...          ...       ...                 ...          ...             ...   
16824    -123.22     39.28                16.0       5569.0          1106.0   
16838    -123.34     39.10                24.0       5372.0          1051.0   
16854    -123.49     38.70                 9.0       5409.0          1019.0   
16859    -123.59     38.80                17.0       5202.0          1037.0   
16986    -124.19     40.73                21.0       5694.0          1056.0   

       population  households  median_income  median_house_value  
0          1015.0       472.0         1.4936             66900.0  
1          1129.0       463.0         1.8200             80100.0  
8          3134.0      1056.0         2.1782             58400.0  
115         890.0       423.0         1.3571             41000.0  
133        4507.0      1807.0         2.2466             66800.0  
...           ...         ...            ...                 ...  
16824      3148.0      1088.0         3.1455            142900.0  
16838      3002.0       992.0         3.0652            131100.0  
16854       594.0       327.0         3.3125            295400.0  
16859      1742.0       803.0         3.1201            176100.0  
16986      2907.0       972.0         3.5363             90100.0  

[1576 rows x 9 columns]

Only 1576 out of 17,000 (somewhere around 9%) areas/blocks/locality have number of bedrooms above 1000

In [35]:
# Logical connectors & | ! can be used to combine multiple conditions
california_housing_dataframe.loc[(california_housing_dataframe.total_bedrooms > 1000) & (california_housing_dataframe.total_rooms > 5000)]

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0        -114.31     34.19                15.0       5612.0          1283.0   
1        -114.47     34.40                19.0       7650.0          1901.0   
133      -116.06     34.15                15.0      10377.0          2331.0   
135      -116.09     34.15                13.0       9444.0          1997.0   
167      -116.24     33.72                25.0       5236.0          1039.0   
...          ...       ...                 ...          ...             ...   
16824    -123.22     39.28                16.0       5569.0          1106.0   
16838    -123.34     39.10                24.0       5372.0          1051.0   
16854    -123.49     38.70                 9.0       5409.0          1019.0   
16859    -123.59     38.80                17.0       5202.0          1037.0   
16986    -124.19     40.73                21.0       5694.0          1056.0   

       population  households  median_income  median_house_value  
0          1015.0       472.0         1.4936             66900.0  
1          1129.0       463.0         1.8200             80100.0  
133        4507.0      1807.0         2.2466             66800.0  
135        4166.0      1482.0         2.6111             65600.0  
167        2725.0       935.0         3.7750             93400.0  
...           ...         ...            ...                 ...  
16824      3148.0      1088.0         3.1455            142900.0  
16838      3002.0       992.0         3.0652            131100.0  
16854       594.0       327.0         3.3125            295400.0  
16859      1742.0       803.0         3.1201            176100.0  
16986      2907.0       972.0         3.5363             90100.0  

[1169 rows x 9 columns]

Some other usefull functions to explore: *isin()*, *isnull()*, *notnull()*, 

## Manipulating Data

In [36]:
cities['Population_in_M'] *= 2 

In [37]:
cities

City_names  Population_in_M Initials
I    New Delhi             37.4       ND
II      Mumbai             47.2        M
III   New York             44.0       NY
IV      London             17.4        L

In [38]:
cities['Area_Square_km'] = list(pd.Series([ 1484, 603.4, 783.8, 1572]))
cities['population_density'] = cities['Population_in_M'] / cities['Area_Square_km']

In [39]:
cities

City_names  Population_in_M Initials  Area_Square_km  population_density
I    New Delhi             37.4       ND          1484.0            0.025202
II      Mumbai             47.2        M           603.4            0.078223
III   New York             44.0       NY           783.8            0.056137
IV      London             17.4        L          1572.0            0.011069

#### Using map()
map takes a single value from a series and returns the manipulates the value. In the end returns a manipulated Series.

In [53]:
population_mean = cities.Population_in_M.mean()
cities['Away_from_mean_population'] = cities.Population_in_M.map(lambda p : p - population_mean)

In [54]:
cities

City_names  Population_in_M Initials  Area_Square_km  population_density  \
I    New Delhi             37.4       ND          1484.0            0.025202   
II      Mumbai             47.2        M           603.4            0.078223   
III   New York             44.0       NY           783.8            0.056137   
IV      London             17.4        L          1572.0            0.011069   

     Away_from_mean_population  
I                          0.9  
II                        10.7  
III                        7.5  
IV                       -19.1

apply() is the equivalent method if we want to transform a whole DataFrame by calling a custom method on each row.

In [55]:
cities['Country'] = ['India', 'India', 'USA', 'UK']
cities

City_names  Population_in_M Initials  Area_Square_km  population_density  \
I    New Delhi             37.4       ND          1484.0            0.025202   
II      Mumbai             47.2        M           603.4            0.078223   
III   New York             44.0       NY           783.8            0.056137   
IV      London             17.4        L          1572.0            0.011069   

     Away_from_mean_population Country  
I                          0.9   India  
II                        10.7   India  
III                        7.5     USA  
IV                       -19.1      UK

In [56]:
cities.Country+'--'+cities.City_names

I      India--New Delhi
II        India--Mumbai
III       USA--New York
IV           UK--London
dtype: object

In [57]:
cities.population_density.sum()

0.17063102893758691

Explore other manipulations as and when needed